indico/web/client/js/custom_elements/ind_date_picker.js

```diff
@@ -6,144 +6,263 @@
 // LICENSE file for more details.
 
 import CustomElementBase from 'indico/custom_elements/_base';
+import {getToday, toDateString, DateRange, OpenDateRange} from 'indico/utils/date';
 import {formatDate} from 'indico/utils/date_format';
 import {createDateParser} from 'indico/utils/date_parser';
+import {getWeekInfoForLocale, getFirstDayOfWeek, getWeekdayNames} from 'indico/utils/l10n';
 import {topBottomPosition} from 'indico/utils/positioning';
+import {setNativeInputValue} from 'indico/utils/react_integration';
 
 import './ind_date_picker.scss';
 
 const DEFAULT_LOCALE = document.documentElement.dataset.canonicalLocale;
-const CALENDAR_CELL_COUNT = 42; // 6 rows, 7 days each
+const KEYBOARD_MOVEMENT = {
+  ArrowRight: 'nextday',
+  ArrowLeft: 'previousday',
+  ArrowDown: 'nextweek',
+  ArrowUp: 'previousweek',
+};
 
 customElements.define(
   'ind-date-picker',
   class extends CustomElementBase {
     // Last value of the internal id used in the widgets
     static lastId = 0;
 
+    static attributes = {
+      min: Date, // Minimum selectable date (inclusive)
+      max: Date, // Maximum selectable date (inclusive)
+      format: String, // Date format
+    };
+
     get value() {
       return this.querySelector('input').value;
     }
 
     setup() {
+      const indDatePicker = this;
       const id = `date-picker-${this.constructor.lastId++}`;
       const input = this.querySelector('input');
       const openCalendarButton = this.querySelector('button');
       const formatDescription = this.querySelector('[data-format]');
       const indCalendar = this.querySelector('ind-calendar');
 
-      const dateFormat = formatDescription.textContent
-        .split(':')[1]
-        .trim()
-        .toLowerCase();
-      const parseDate = createDateParser(dateFormat);
+      const parseDate = createDateParser(this.format);
 
-      indCalendar.format = dateFormat;
-      indCalendar.value = toDateString(parseDate(this.value));
+      indCalendar.format = this.format;
+      indCalendar.rangeStart = indCalendar.rangeEnd = toDateString(parseDate(this.value));
+      indCalendar.min = this.min;
+      indCalendar.max = this.max;
       formatDescription.id = `${id}-format`;
       input.setAttribute('aria-describedby', formatDescription.id);
-      openCalendarButton.setAttribute('aria-haspopup', 'dialog');
-      openCalendarButton.setAttribute('aria-controls', true);
 
       // This property is defined here rather than in the class because it
       // relies on the parseDate() function created in this scope.
       Object.defineProperty(this, 'date', {
         get: () => parseDate(input.value),
       });
 
-      openCalendarButton.addEventListener('click', () => {
-        openCalendarButton.disabled = true;
-        requestAnimationFrame(openDialog);
-      });
+      openCalendarButton.addEventListener('click', open);
+      input.addEventListener('click', open);
```

Comment:
Is this a good idea? At least as someone who often prefers typing over selecting via UI, it's now very awkward to do that - especially for the normal date field.

In case of the date range field, I think it makes a bit more sense because you have to pick two dates and see the ranges that you can select from.

If it's better usability-wise to open it regardless of where in the field you click, I think we should at least not close-and-reopen when you click the field again while the picker is open. Instead, just closing it would be nice because then you can just click a second time in order to type a date manually.

Reply:
Hmm... yeah. I guess it was a bad move. I'll fix it.